File: haskell/glue/src/Glue/Lib/List/Concat.hs
module Glue.Lib.List.Concat where

import Glue.Eval (Eval, throwError)
import Glue.Eval.Exception (wrongArgumentType)
import Glue.IR (IR (..))

concat :: IR Eval
concat = NativeFunc concatImpl

concatImpl :: IR Eval -> Eval (IR Eval)
concatImpl list1 = pure $ NativeFunc (concatTo list1)

concatTo :: IR Eval -> IR Eval -> Eval (IR Eval)
concatTo list1 list2 = case (list1, list2) of
    (List xs, List ys) -> pure $ List (xs <> ys)
    _ -> throwError $ wrongArgumentType ["list", "list"]

File: dart/glue/lib/src/lib/list/concat.dart
import 'package:glue/src/eval.dart';
import 'package:glue/src/eval/exception.dart';
import 'package:glue/src/ir.dart';

/// Concat function - concatenates two lists
/// Mirrors Haskell Glue.Lib.List.Concat.concat exactly
Ir concat = IrNativeFunc(concatImpl);

/// Concat function implementation
/// Mirrors Haskell Glue.Lib.List.Concat.concatImpl exactly
Eval<Ir> concatImpl(Ir list1) {
  return Eval.pure(IrNativeFunc(concatWith(list1)));
}

/// Helper function for second list argument
/// Mirrors Haskell Glue.Lib.List.Concat.concatWith exactly
Eval<Ir> Function(Ir) concatWith(Ir list1) {
  return (Ir list2) => switch ((list1, list2)) {
    (IrList(elements: final e1), IrList(elements: final e2)) => Eval.pure(
      IrList([...e1, ...e2]),
    ),
    _ => throwError(wrongArgumentType(['list', 'list'])),
  };
}
